source/tutorials/learning-journey/packaging-existing-software.md

```diff
@@ -0,0 +1,531 @@
+       >   195 | # warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE"
+       >       |   ^~~~~~~
+       > icat.c:39:10: fatal error: Imlib2.h: No such file or directory
+       >    39 | #include <Imlib2.h>
+       >       |          ^~~~~~~~~~
+       > compilation terminated.
+       > make: *** [Makefile:16: icat.o] Error 1
+       For full logs, run 'nix log /nix/store/l5wz9inkvkf0qhl8kpl39vpg2xfm2qpy-icat.drv'.
+```
+
+A compiler error! The `icat` source was pulled from GitHub, and Nix tried to build what it found, but compilation failed due to a missing dependency: the `imlib2` header.
+
+If you [search for `imlib2` on search.nixos.org](https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=imlib2), you'll find that `imlib2` is already in `nixpkgs`.
+
+You can add this package to your build environment by adding `imlib2` to the set of inputs to the expression in `icat.nix`, and then adding `imlib2` to the list of `buildInputs` in `stdenv.mkDerivation`:
+
+```nix
+# icat.nix
+{ lib
+, stdenv
+, fetchFromGitHub
+, imlib2
+}:
+
+stdenv.mkDerivation {
+  name = "icat";
+
+  src = fetchFromGitHub {
+    owner = "atextor";
+    repo = "icat";
+    rev = "v0.5";
+    sha256 = "0wyy2ksxp95vnh71ybj1bbmqd5ggp13x3mk37pzr99ljs9awy8ka";
+  };
+
+  buildInputs = [ imlib2 ];
+}
+```
+
+Another error, but compilation proceeds further this time:
+
+```console
+$ nix-build -A icat
+this derivation will be built:
+  /nix/store/bw2d4rp2k1l5rg49hds199ma2mz36x47-icat.drv
+...
+error: builder for '/nix/store/bw2d4rp2k1l5rg49hds199ma2mz36x47-icat.drv' failed with exit code 2;
+       last 10 log lines:
+       >                  from icat.c:31:
+       > /nix/store/hkj250rjsvxcbr31fr1v81cv88cdfp4l-glibc-2.37-8-dev/include/features.h:195:3: warning: #warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE" [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wcpp-Wcpp8;;]
+       >   195 | # warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE"
+       >       |   ^~~~~~~
+       > In file included from icat.c:39:
+       > /nix/store/4fvrh0sjc8sbkbqda7dfsh7q0gxmnh9p-imlib2-1.11.1-dev/include/Imlib2.h:45:10: fatal error: X11/Xlib.h: No such file or directory
+       >    45 | #include <X11/Xlib.h>
+       >       |          ^~~~~~~~~~~~
+       > compilation terminated.
+       > make: *** [Makefile:16: icat.o] Error 1
+       For full logs, run 'nix log /nix/store/bw2d4rp2k1l5rg49hds199ma2mz36x47-icat.drv'.
+```
+
+You can see a few warnings which should be corrected in the upstream code, but the important bit for this tutorial is `fatal error: X11/Xlib.h: No such file or directory`: another dependency is missing.
+
+
+:::{note}
+Determining from where to source a dependency is currently a somewhat-involved process: it helps to become familiar with searching the `nixpkgs` source for keywords.
```

Comment:
This doesn't actually help a user solve their problem, they're almost guaranteed not to be familiar with `nixpkgs` source. I would recommend to instead tell them how to search `nixpkgs` or whatever it is they _actually_ need to do.

Reply:
Agreed, but developing those heuristics is probably another whole tutorial, which we probably could just point to from this note once it exists (if)

Reply:
Yeah I think this would be better as a separate tutorial, it's really quite involved (grep, `unsafeGetAttrPos`, `nix edit`, the structure of Nixpkgs, backtracking through definitions)

Reply:
Discussed online: mention `nix-locate`, we don't mind that it's a third-party tool